Library/Homebrew/extend/git_repository.rb

```diff
@@ -7,41 +7,57 @@ def git?
   end
 
   def git_origin
-    return unless git? && Utils.git_available?
-    cd do
-      Utils.popen_read("git", "config", "--get", "remote.origin.url").chuzzle
-    end
+    git_client.run("config", "--get", "remote.origin.url")
   end
 
   def git_head
-    return unless git? && Utils.git_available?
-    cd do
-      Utils.popen_read("git", "rev-parse", "--verify", "-q", "HEAD").chuzzle
-    end
+    git_client.run("rev-parse", "--verify", "-q", "HEAD")
   end
 
   def git_short_head
-    return unless git? && Utils.git_available?
-    cd do
-      Utils.popen_read(
-        "git", "rev-parse", "--short=4", "--verify", "-q", "HEAD"
-      ).chuzzle
-    end
+    git_client.run("rev-parse", "--short=4", "--verify", "-q", "HEAD")
   end
 
   def git_last_commit
-    return unless git? && Utils.git_available?
-    cd do
-      Utils.popen_read("git", "show", "-s", "--format=%cr", "HEAD").chuzzle
-    end
+    git_client.run("show", "-s", "--format=%cr", "HEAD")
+  end
+
+  def git_branch
+    git_client.run("rev-parse", "--abbrev-ref", "HEAD")
   end
 
   def git_last_commit_date
-    return unless git? && Utils.git_available?
-    cd do
-      Utils.popen_read(
-        "git", "show", "-s", "--format=%cd", "--date=short", "HEAD"
-      ).chuzzle
+    git_client.run("show", "-s", "--format=%cd", "--date=short", "HEAD")
+  end
+
+  private
+
+  def git_client
+    @client ||= begin
+      if git? && Utils.git_available?
+        GitClient.new(self)
+      else
+        NullClient.new
+      end
     end
   end
+
+  class GitClient
+    attr_reader :path
+
+    def initialize(path)
+      @path = path
+    end
+
+    def run(*args)
+      raise TapUnavailableError, path.name unless path.directory?
```

Comment:
Posting again just because it's hidden using the "outdated diff" feature:

Maybe this should be raised on initialization? Or perhaps we can include this in the `git_client` method, and then use a NullClient when it's not available?

I'll address the related test failure once I've got a better idea of when I wanna raise this

Reply:
I think this is the right place to raise, but not the right exception, since the path doesn‘t have to be a Tap.

Reply:
True. Although in the context that this is used in `tap.rb`, it seems like `TapUnavailableError` would make the most sense. What do you think about passing an `on_error` block to the constructor?

Reply:
> What do you think about passing an on_error block to the constructor?

Give it a go and let's see what it looks like, thanks!